plans/data-transfer-seeds-leeches/utils/createnode.go

```diff
@@ -0,0 +1,155 @@
+	"github.com/ipfs/go-unixfs/importer/helpers"
+	"github.com/ipfs/go-unixfs/importer/trickle"
+	"github.com/ipfs/testground/sdk/runtime"
+	libp2p "github.com/libp2p/go-libp2p"
+	host "github.com/libp2p/go-libp2p-host"
+	"github.com/multiformats/go-multihash"
+	"github.com/pkg/errors"
+)
+
+type Node struct {
+	Host    host.Host
+	Bitswap *bs.Bitswap
+	Dserv   ipld.DAGService
+}
+
+func (n *Node) Close() {
+	n.Bitswap.Close()
+}
+
+// CreateNode creates a libp2p Node with a Bitswap instance
+func CreateNode(ctx context.Context, runenv *runtime.RunEnv) (*Node, error) {
+	bstoreDelay := runenv.IntParamD("bstore_delay_ms", 0)
+
+	h, err := libp2p.New(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	routing, err := nilrouting.ConstructNilRouting(nil, nil, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+	net := bsnet.NewFromIpfsHost(h, routing)
+
+	bsdelay := delay.Fixed(time.Duration(bstoreDelay) * time.Millisecond)
+	dstore := ds_sync.MutexWrap(delayed.New(ds.NewMapDatastore(), bsdelay))
+	bstore, err := blockstore.CachedBlockstore(ctx,
+		blockstore.NewBlockstore(ds_sync.MutexWrap(dstore)),
+		blockstore.DefaultCacheOpts())
+	if err != nil {
+		return nil, err
+	}
+
+	bitswap := bs.New(ctx, net, bstore).(*bs.Bitswap)
+	bserv := blockservice.New(bstore, bitswap)
+	dserv := merkledag.NewDAGService(bserv)
+	return &Node{h, bitswap, dserv}, nil
+}
+
+type AddSettings struct {
+	Layout    string
+	Chunker   string
+	RawLeaves bool
+	Hidden    bool
+	NoCopy    bool
+	HashFunc  string
+	MaxLinks  int
+}
+
+func (n *Node) Add(ctx context.Context, r io.Reader) (ipld.Node, error) {
+	settings := AddSettings{
+		Layout:    "balanced",
+		Chunker:   "size-262144",
+		RawLeaves: false,
+		Hidden:    false,
+		NoCopy:    false,
+		HashFunc:  "sha2-256",
+		MaxLinks:  helpers.DefaultLinksPerBlock,
+	}
+	// for _, opt := range opts {
+	// 	err := opt(&settings)
+	// 	if err != nil {
+	// 		return nil, err
+	// 	}
+	// }
+
+	prefix, err := merkledag.PrefixForCidVersion(1)
+	if err != nil {
+		return nil, errors.Wrap(err, "unrecognized CID version")
+	}
+
+	hashFuncCode, ok := multihash.Names[strings.ToLower(settings.HashFunc)]
+	if !ok {
+		return nil, errors.Wrapf(err, "unrecognized hash function %q", settings.HashFunc)
+	}
+	prefix.MhType = hashFuncCode
+
+	dbp := helpers.DagBuilderParams{
+		Dagserv:    n.Dserv,
+		RawLeaves:  settings.RawLeaves,
+		Maxlinks:   settings.MaxLinks,
+		NoCopy:     settings.NoCopy,
+		CidBuilder: &prefix,
+	}
+
+	chnk, err := chunker.FromString(r, settings.Chunker)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create chunker")
+	}
+
+	dbh, err := dbp.New(chnk)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create dag builder")
+	}
+
+	var nd ipld.Node
+	switch settings.Layout {
+	case "trickle":
+		nd, err = trickle.Layout(dbh)
+	case "balanced":
+		nd, err = balanced.Layout(dbh)
+	default:
+		return nil, errors.Errorf("unrecognized layout %q", settings.Layout)
+	}
+
+	return nd, err
+}
```

Comment:
Oh cool! This func makes using ipfs.add way simpler. I had to learn how to convert files to `go-ipfs-files` File structs so that I could add them to IPFS, resulting in https://github.com/ipfs/testground/pull/58/files#diff-f3dda0462ce5f0f7e66209ffcbdf9666 and https://github.com/ipfs/go-ipfs/tree/master/docs/examples/go-ipfs-as-a-library

@Stebalien should we port this func into the https://github.com/ipfs/interface-go-ipfs-core ?

Reply:
I ported some of this code from the excellent p2plab by @hinshun

We may not need all these configurable options but I've left them in there so we can easily enable them if we want to see what happens with eg different chunkers

Reply:
This should _probably_ live in `go-unixfs`. However, I'm not sure how to shoehorn it in.

That is, `go-unixfs` should have:

```
func Import(dag DAGService, node files.Node, Options...) (cid.Cid, error)
func Export(dag DAGService, cid.Cid) (files.Node, error)
```

(and it should depend on interface-go-ipfs-core).